### DISCUSSION TITLE
After clicking on a link and returning to the previous page, it redirects to the link page if the mouse hovers the link

### DISCUSSION THREAD

**Discussion by damsdev1** [vercel/next.js#50958] [Help]
### Summary

I don't understand this bug, I think it's directly linked to NextJS, but I'd like to be sure. This bug occurs when a link doesn't go to the same layout, that's what I've deduced.

Here's a demonstration of the bug (when I click, the cursor displays yellow), you can see that just a mouse-over on the link without clicking simulates a click.
![bug](https://github.com/vercel/next.js/assets/60252259/c9ddfe0a-df59-4338-a233-2de44cc31092)

This bug doesn't appear in development mode, just in production mode. This bug doesn't appear also if link prefetching is disabled.

To reproduce the bug, you need to open the link in a new tab and not use the codesandbox preview.

### Example

https://codesandbox.io/p/sandbox/zen-tdd-ukk3pe

**Comment by jessicaribeiroalves** [comment]
I am facing the same issue on my website. In my case I don't have a different layout, disabling the prefetch solved the issue, but now chrome has two entries of the same page in the browser history. Everything works as expected in safari.

**Reply by sr-tamim** [reply]
Bug doesn't appear on Edge either. It only appears on Chrome

**Comment by damsdev1** [comment]
> I am facing the same issue on my website. In my case I don't have a different layout, disabling the prefetch solved the issue, but now chrome has two entries of the same page in the browser history. Everything works as expected in safari.

Everything works on Firefox also.
This bug seems to be linked to Chromium-based browsers

**Reply by jessicaribeiroalves** [reply]
I just removed the `generateStaticParams` from the page and there's no duplicate pages on history browser anymore.

**Reply by sr-tamim** [reply]
Bug doesn't appear on Edge either. It only appears on Chrome

**Comment by icyJoseph** [comment]
At first I thought that this was somehow, History Manipulation at its finest, https://bugs.chromium.org/p/chromium/issues/detail?id=907167, but the second report makes doubt...

@DamsDev1 could you please, put the demo in a github repository?

**Reply by damsdev1** [reply]
> At first I thought that this was somehow, History Manipulation at its finest, https://bugs.chromium.org/p/chromium/issues/detail?id=907167, but the second report makes doubt...
> 
> @DamsDev1 could you please, put the demo in a github repository?

Yes, sure: https://github.com/DamsDev1/nextjs-hovering-bug

**Reply by icyJoseph** [reply]
What am I supposed to do? I click on `test` that takes me to `test2`, then click back on the browser, and then I hover on the `test` link but nothing happens.

Could you perhaps have done a permanent redirect or something like that? Do you see the issue in Incognito Mode?

**Reply by icyJoseph** [reply]
After I tested your project, the next project I tried, was getting redirect to `/test` 🤔 had to hard reload the page

**Reply by damsdev1** [reply]
> What am I supposed to do? I click on `test` that takes me to `test2`, then click back on the browser, and then I hover on the `test` link but nothing happens.
> 
> Could you perhaps have done a permanent redirect or something like that? Do you see the issue in Incognito Mode?

Do you have run it in production mode ?

**Reply by icyJoseph** [reply]
I did, but you know what I went back, and noticed once thing. There's a special combination that leads to this. If I click somewhere in `/test` page, it works normally, if I quickly just click on it, and then navigate back, on hover, I navigate to `/test2`... 

- https://web.dev/bfcache/

However, it goes on and off, I cannot reliably do it. I wanted to shoot a video, and now it works fine again... Perhaps the bfcache is playing us here?

**Reply by icyJoseph** [reply]
My take is that `redirect from `/` to `/test`, and then, with no user interactions on `/test`, other than clicking the link to `/test2`, tells Chrome that the `/test` page is "skippable".

**Reply by icyJoseph** [reply]
Try cleaning your cache the hard way, so that it clears the permanent redirect, and this time, interact (click the whitespace) of `/test`.

**Reply by damsdev1** [reply]
> I did, but you know what I went back, and noticed once thing. There's a special combination that leads to this. If I click somewhere in `/test` page, it works normally, if I quickly just click on it, and then navigate back, on hover, I navigate to `/test2`...
> 
> * https://web.dev/bfcache/
> 
> However, it goes on and off, I cannot reliably do it. I wanted to shoot a video, and now it works fine again... Perhaps the bfcache is playing us here?

Bfcache is not to blame for this problem. It does play a part in the problem, because if I disable it, this bug no longer occurs. 
However, if I disable JavaScript code execution in the debugger, the bug no longer occurs either.

This bug is due to the **prefetching of the Link component**.

Prefetching occurs either when the mouse is moved **over the link**, or when the **link becomes visible in the viewport**. 
I've noticed that if you put two links, one already in the viewport and another outside the viewport, when you return to the page containing the two links with the bfcache, if you scroll and display the link that wasn't yet in the viewport, the prefetching redirects to the old link that has already been clicked.

https://github.com/vercel/next.js/assets/60252259/5574a295-79e7-490b-bf01-27f7b969cf86

Test website: https://ukk3pe-8080.csb.app/test

**Reply by icyJoseph** [reply]
That codesandbox demo never loaded for me, it hangs forever.

Anyway, I know now how to trigger it 100% of the time! I have to load the page, and never open the devTools, or open them before I even go to localhost:3000, otherwise, it counts as an interaction itself, and the page is not bfcache'd. 

It is a tricky damn bug, proper of the History Manipulation kind!

https://github.com/vercel/next.js/assets/21013447/a78304ee-995c-4216-aef3-0285a6dcd74a

I'll try to dig up more info on this matter, but that BFCache is for sure the root of many headaches

**Reply by damsdev1** [reply]
> That codesandbox demo never loaded for me, it hangs forever.

https://codesandbox.io/p/sandbox/zen-tdd-ukk3pe maybe with full source code ?

**Reply by jessicaribeiroalves** [reply]
@icyJoseph if you go to the deployed version that @DamsDev1 just shared you'll notice the weird behaviour triggered by hovering or showing the second link on the viewport.
Here are the steps to reproduce the bug:
- Go to https://ukk3pe-8080.csb.app/test
- Click on test link
- Press the back button and go to the previous page (test)
- If you hover the test link or scroll to the bottom of the page where there's a second link, you'll be thrown to the test 2 page (which is the one you clicked before)

**Reply by icyJoseph** [reply]
@jessicaribeiroalves the problem was that this bug needs the `/test` page to be labelled as non-interactive, and by opening the devTools I was breaking that. I've managed now to do it myself. 

~My impression is that the History API is not triggering the "popstate" event, and that's the root of all evil. I'll try to make a quick test for it.~

~Or that the JS context is not truly destroyed when clicking the back button. Confusing stuff.~

**Reply by icyJoseph** [reply]
Nope, I'm wrong here, that's prefetching test2 so it's correct.

**Reply by icyJoseph** [reply]
https://github.com/vercel/next.js/assets/21013447/b3dbadf9-c6dc-47af-8fc1-151322da2ddf

This is a true madness rabbit hole, but side by side, left is with the BFCache on test, right side is without it.

One decides is time to `navigate`, the other decides it is time to `prefetch`, as you'd expect from a Prefetching activity...

**Reply by damsdev1** [reply]
> Screen.Recording.2023-06-12.at.19.20.37.mov 
> This is a true madness rabbit hole, but side by side, left is with the BFCache on test, right side is without it.
> 
> One decides is time to `navigate`, the other decides it is time to `prefetch`, as you'd expect from a Prefetching activity...

I managed to find the source of the problem. When a RootLayout is added to the root without using groups, this bug disappears. However, it shouldn't appear because it's possible to make RootLayouts for each group. https://nextjs.org/docs/app/building-your-application/routing/route-groups#creating-multiple-root-layouts

However, when creating a RootLayout, a new bug arises: style collision, where styles from different layouts are mixed when they're not the same.
Code sandbox: https://codesandbox.io/p/sandbox/wizardly-khorana-sy3j95
Test url: https://sy3j95-3000.csb.app/


https://github.com/vercel/next.js/assets/60252259/935eb9d8-7829-46b2-859c-7d0b566434f5

**Reply by damsdev1** [reply]
Any updates ?

**Reply by soyuka** [reply]
possibly reproduced at https://github.com/vercel/next.js/issues/51565

**Reply by icyJoseph** [reply]
Sorry y'all, got very busy this week and now I'm Out of Office - that issue looks about the same though.

**Reply by alexisdrakopoulos** [reply]
I ended up switching to amplify

**Comment by danielpopperl** [comment]
same bug here.

next v13.4.4

**Reply by stx-chris** [reply]
same here, v13.4.12. occurs only in Chrome

**Comment by danielpopperl** [comment]
@DamsDev1 where did you depoy your site?
I deployed on Netlify and I was facing same bug, but now I deploy on Vercel and this bug disappear. For somehow Netlify has other "engine" of deploy and its not compatible with new Nextjs/React

**Reply by jessicaribeiroalves** [reply]
yeah that's not the first issue I face deploying a Next application on Netlify using the last nextjs version, unfortunately :'(

**Reply by alexisdrakopoulos** [reply]
I'm also on netlify having the same issue, I'll try deploying elsewhere.

**Comment by elisaado** [comment]
Current workaround we use is to detect links that go outside Next, and use a plain `<a>` tag for them. Any links that point inside the same app still use `<Link>`.

**Comment by abdulimr** [comment]
Also, something we're experiencing when the elements are ```<Link>```s.

Disabling `prefetch` using `prefetch={false}` solves the problem, though of course, disabled prefetching - not an ideal fix and defeats the purpose of using the ```Link``` component in the first place.

Note that this occurred when building for a static export and hosting on AWS S3.

**Comment by yannicka** [comment]
I have the exact same issue on Firefox 117 with Next.js 13.4.10 (and using the "/app" folder).

**Reply by alexisdrakopoulos** [reply]
To anyone who encounters this issue, if you're using netlify and this issue doesn't occur locally try deploying elsewhere. I switched to vercel and the issue went away

**Reply by yannicka** [reply]
For my part I have only tested locally for the moment.

**Reply by Narixius** [reply]
> To anyone who encounters this issue, if you're using netlify and this issue doesn't occur locally try deploying elsewhere. I switched to vercel and the issue went away

Hey @alexisdrakopoulos
I am facing this issue on Vercel!

**Reply by brunocabral88** [reply]
I'm facing the same issue on a static export of NextJS 14.1.3 on Chrome. Current NextJS implementation of `<Link />` doesn't seem to play well with browsers bfcache. Ended up resorting to `<a>` tags for external links.

**Comment by KirillPirinen** [comment]
Are there any updates on this? Have app using 14+ nextjs and this problem still here. Even more, it also appears when I disable prefetch, or using .push instead of link. (

**Comment by Ahmadnoorkhan1** [comment]
I am also facing this issue on "next": "14.2.3", i set prefetch false on the Link tags and now i am not facing the issue. Yet my links are being loading really slow. Is there an explanation for that????
